node/core/approval-voting/src/tests.rs

```diff
@@ -1280,3 +1703,858 @@ fn forkful_import_at_same_height_act_on_leaf() {
+	let (mock_clock, _) = make_harness_input();
+	let store = SharedTestStore::new(TestStore::default());
+	let assignment_criteria = MockAssignmentCriteria(|| {
+			let mut assignments = HashMap::new();
+			let _ = assignments.insert(CoreIndex(0), approval_db::v1::OurAssignment {
+				cert: garbage_assignment_cert(
+					AssignmentCertKind::RelayVRFModulo { sample: 0 }
+				),
+				tranche: 0,
+				validator_index: ValidatorIndex(0),
+				triggered: false,
+			}.into());
+			assignments
+		},
+		|_| Ok(0),
+	);
+
+	test_harness(Box::new(oracle), mock_clock, store.clone(), assignment_criteria, |test_harness| async move {
+		let TestHarness {
+			mut virtual_overseer,
+			clock,
+			..
+		} = test_harness;
+
+		let block_hash = Hash::repeat_byte(0x01);
+		let candidate_receipt = CandidateReceipt::<Hash>::default();
+		let candidate_hash = candidate_receipt.hash();
+		let slot = Slot::from(1);
+		let candidate_index = 0;
+		let validator0 = ValidatorIndex(0);
+
+		let validators = vec![
+			Sr25519Keyring::Alice,
+			Sr25519Keyring::Bob,
+			Sr25519Keyring::Charlie,
+			Sr25519Keyring::Dave,
+			Sr25519Keyring::Eve,
+		];
+		let session_info = SessionInfo {
+			validators: validators.iter().map(|v| v.public().into()).collect(),
+			validator_groups: vec![
+				vec![ValidatorIndex(0), ValidatorIndex(1)],
+				vec![ValidatorIndex(2)],
+				vec![ValidatorIndex(3), ValidatorIndex(4)],
+			],
+			needed_approvals: 2,
+			discovery_keys: validators.iter().map(|v| v.public().into()).collect(),
+			assignment_keys: validators.iter().map(|v| v.public().into()).collect(),
+			n_cores: validators.len() as _,
+			zeroth_delay_tranche_width: 5,
+			relay_vrf_modulo_samples: 3,
+			n_delay_tranches: 50,
+			no_show_slots: 2,
+		};
+
+		ChainBuilder::new()
+			.add_block(block_hash, ChainBuilder::GENESIS_HASH, 1, BlockConfig {
+				slot,
+				candidates: Some(vec![(candidate_receipt, CoreIndex(0), GroupIndex(0))]),
+				session_info: Some(session_info),
+			})
+			.build(&mut virtual_overseer).await;
+
+		assert!(!clock.inner.lock().has_wakeup(1));
+		clock.inner.lock().wakeup_all(1);
+
+		let rx = check_and_import_assignment(
+			&mut virtual_overseer,
+			block_hash,
+			candidate_index,
+			validator0,
+		).await;
+		assert_eq!(rx.await, Ok(AssignmentCheckResult::Accepted));
+
+		assert!(
+			should_trigger_assignment(
+				&mut virtual_overseer, store, clock, slot, 0, block_hash, candidate_hash, candidate_index,
+			).await
+		);
+
+		virtual_overseer
+	});
+}
+
+async fn should_trigger_assignment(
+	virtual_overseer: &mut VirtualOverseer,
+	store: impl Backend,
+	clock: Box<MockClock>,
+	slot: Slot,
+	tranche_offset: Tick,
+	block_hash: Hash,
+	candidate_hash: CandidateHash,
+	candidate_index: CandidateIndex,
+) -> bool {
+		assert!(clock.inner.lock().has_wakeup(slot_to_tick(slot) + tranche_offset));
+		clock.inner.lock().wakeup_all(slot_to_tick(slot) + tranche_offset);
+
+		async_std::task::sleep(std::time::Duration::from_millis(200)).await;
```

Comment:
I'd prefer to craft explicit features,i.e. a oneshot with a timeout to assure synchronization _if_ possible.